### PR TITLE
feat(pkg/reconciler) : Adding a reconciler in OSM for the CA bundle in the mutating webhook configuration

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

This PR added a controller that watched for mutating webhook resources in OSM and updates only OSM's mutating webhook with the latest CA bundle, whenever it is missing ( due to deletion) 

Fixes : #1909 

**Affected area**:

- New Functionality      [X]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

File: cmd/osm-controller/osm-controller.go [hunk 1]
```diff
@@ -98,6 +110,15 @@ func init() {
...
 	flags.BoolVar(&optionalFeatures.Backpressure, "enable-backpressure-experimental", false, "Enable experimental backpressure feature")
+
+	// k8s controller manager options
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
+
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = v1beta1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:121]: Could you describe as code comments what controller manager is, what metrics address means here, and so on.

File: cmd/osm-controller/osm-controller.go [hunk 2]
```diff
@@ -229,6 +250,43 @@ func main() {
...
+		os.Exit(1)
+	}
+	// +kubebuilder:scaffold:builder
+
+	log.Info().Msg("starting manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		log.Error().Err(err).Msg("problem running manager")
+		os.Exit(1)
+	}
+	log.Info().Msg("Successfully running controller manager")
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:288]: Would be better to move this initialization to the a separate function.

File: cmd/osm-controller/osm-controller.go [hunk 3]
```diff
@@ -229,6 +250,43 @@ func main() {
...
+		LeaderElection:     enableLeaderElection,
+		Port:               9443,
+		Namespace:          osmNamespace,
+	})
+	if err != nil {
+		log.Error().Err(err).Msg("Error starting up controller manager")
+		os.Exit(1)
+	}
+
+	log.Info().Msg("Successfully setup controller for reconcile")
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:267]:
```suggestion
	log.Info().Msg("Successfully setup controller for resource reconciliation")
```

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
+package reconciler
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:1]: Would it make sense to rename this file to `reconcile.go`? This could be a pattern we follow for the different resource reconciliation, ex: `reconciler/<resource>/reconcile.go`. If the same file name cannot be reused within a package, we could keep it as is.

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/logger"
+)
+
+var log = logger.New("mutatingwebhook-reconciler")
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:18]: I'd create a logger for the reconciler package to be consistent with existing logging facility, where we use package scoped loggers.

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+type MutatingWebhookConfigrationReconciler struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	OsmWebhook   string
+	OsmNamespace string
+	CertManager  certificate.Manager
+}
+
+// +kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=mutatingwebhookconfigurations,verbs=get;list;watch;update;
+// +kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=mutatingwebhookconfigurations/status,verbs=get;update;patch
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:30]: Why is this required?

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// reconcile only for OSM mutatingWebhookConfiguration
+	var shouldUpdate bool
+	if r.OsmWebhook == instance.Name {
+		//check if CA bundle exists on webhook
+		for idx, webhook := range instance.Webhooks {
+			// CA bundle missing for webhook, update webhook to include the latest CA bundle
+			if webhook.Name == "osm-inject.k8s.io" && webhook.ClientConfig.CABundle == nil {
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:48]: Could you make `mutatingWebhookName` an exported constant in the `webhook.go` file and use it as `injector.MutatingWebhookName` here.

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+				log.Trace().Msgf("CA bundle missing for webhook : %s ", req.Name)
+				shouldUpdate = true
+				cn := certificate.CommonName(fmt.Sprintf("%s.%s.svc", constants.OSMControllerName, r.OsmNamespace))
+				cert, err := r.CertManager.GetCertificate(cn)
+				if err != nil {
+					return ctrl.Result{}, errors.Errorf("Unable to update mutating webhook, Error getting certificate for the mutating webhook: %+v", err)
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:54]: Could you augment the log message to something as:
```suggestion
					return ctrl.Result{}, errors.Errorf("Error updating mutating webhook, unable to get certificate for the mutating webhook %s: %s" webhookName, err)
```

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+				}
+				instance.Webhooks[idx].ClientConfig.CABundle = cert.GetCertificateChain()
+			}
+		}
+	}
+
+	if !shouldUpdate {
+		log.Trace().Msgf("Mutating webhook configuration already compliant %s", req.Name)
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:62]: Please log the name of the mutatingwebhookconfiguration.

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Update(ctx, instance); err != nil {
+		log.Error().Err(err).Msgf("failure reading object %s", req.NamespacedName.String())
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:67]:
```suggestion
		log.Error().Err(err).Msgf("Error updating mutatingwebhookconfiguration %s: %s", webhookConfigName, err)
```

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	log.Trace().Msgf("Successfully updated mutating webhook configuration CA bundle for : %s ", req.Name)
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:71]:
```suggestion
	log.Trace().Msgf("Successfully updated mutating webhook configuration CA bundle for mutatingwebhookconfiguration %s ", req.Name)
```

Is `req.Name` the name of the webhook config?

File: pkg/reconciler/mutatingwebhook/suite_test.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = BeforeSuite(func(done Done) {
+	logf.SetLogger(zap.New())
+
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths: []string{filepath.Join("..", "config", "crd", "bases")},
+	}
```
Comment (shashankram) [inline pkg/reconciler/mutatingwebhook/suite_test.go:40]: What's happening here? Is this specific to your environment?

Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/suite_test.go:40]: This is specific to the kubernetes test environment to start/stop a kubernetes control plane for testing the controllers

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 11 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:71]: yes `req.Name` is the name for the requesting mutatingwebhook i.e. the webhook which triggered the reconciler

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 9 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:62]: it is already being logged with `req.Name`

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 5 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:18]: I've updated the logger to be package scoped

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 6 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:30]: we are using kubebuilder to test the controllers, and these lines add the necessary permissions in kubebuilders to listen on events related to muutatingwebhooks

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 4 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go:1]: seems like it worked, so I have updated the naming

File: cmd/osm-controller/osm-controller.go (hunk 2 above)
Comment (snehachhabria) [reply cmd/osm-controller/osm-controller.go:288]: done

File: cmd/osm-controller/osm-controller.go (hunk 1 above)
Comment (snehachhabria) [reply cmd/osm-controller/osm-controller.go:121]: added some comments

File: pkg/reconciler/mutatingwebhook/suite_test.go (hunk 12 above)
Comment (shashankram) [reply pkg/reconciler/mutatingwebhook/suite_test.go:40]: What does kubernetes control plane refer to here?

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 6 above)
Comment (shashankram) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go]: I don't understand this since I am not familiar with kubebuilder, but if you think a few comments on this could help readability they would be welcome.

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 11 above)
Comment (shashankram) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go]: ok, I'd mention the resource name as well as suggested in the comment.

File: pkg/reconciler/mutatingwebhook/suite_test.go (hunk 12 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/suite_test.go:40]: Check these two docs : https://book-v1.book.kubebuilder.io/beyond_basics/running_tests.html , https://book.kubebuilder.io/reference/envtest.html. Do you think it would be good to reference these in the code as comments so that everyone can use it as a reference ?

File: pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go (hunk 6 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/mutatingwebhook_controller.go]: I'm not sure what comments I can add, but do you think a reference to a doc would help ?

**Review by shashankram — APPROVED** [review]

File: pkg/reconciler/mutatingwebhook/reconcile.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+	instance := &v1beta1.MutatingWebhookConfiguration{}
+
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		log.Error().Err(err).Msgf("failure reading object %s ", req.NamespacedName.String())
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// reconcile only for OSM mutatingWebhookConfiguration
+	var shouldUpdate bool
+	if r.OsmWebhook == instance.Name {
```
Comment (ritazh) [inline pkg/reconciler/mutatingwebhook/reconcile.go:45]: Should we check for if `req.NamespacedName` is OSM mutatingWebhookConfig at the beginning of this function?

File: pkg/reconciler/mutatingwebhook/reconcile.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	if !shouldUpdate {
+		log.Trace().Msgf("Mutatingwebhookconfiguration %s already compliant", req.Name)
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Update(ctx, instance); err != nil {
+		log.Error().Err(err).Msgf("Error updating mutatingwebhookconfiguration %s: %s", req.Name, err)
+		return ctrl.Result{Requeue: false}, nil
```
Comment (ritazh) [inline pkg/reconciler/mutatingwebhook/reconcile.go:69]: If the update fails, do we not want to retry?

File: cmd/osm-controller/osm-controller.go [hunk 15]
```diff
@@ -353,3 +383,44 @@ func getOSMControllerPod(kubeClient kubernetes.Interface) (*corev1.Pod, error) {
...
 }
+
+//Setting up k8s controller manager to reconcile OSM resources
+func createControllerManagerForOSMResources(certManager certificate.Manager) error {
+	log.Info().Msg("Setting up controller manager to reconcile OSM resources")
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: metricsAddr,
+		LeaderElection:     enableLeaderElection,
+		Port:               9443,
```
Comment (ritazh) [inline cmd/osm-controller/osm-controller.go:394]: Do you need `Port`?

Comment (snehachhabria) [reply cmd/osm-controller/osm-controller.go]: That's right, I have removed it

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 13 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/reconcile.go:45]: I thought a mutatingwebhookconfiguration is namespace independent and hence checking against the OSM namespace would be redundant for this particular scenario

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 14 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/reconcile.go:69]: We could update to retry, and that would help for transient issues but it there's an actual issue then we would be stuck in a reconcile loop. WDYT ?

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 13 above)
Comment (ritazh) [reply pkg/reconciler/mutatingwebhook/reconcile.go:45]: I was suggesting that we check the request mutatingwebhookconfig's name `req.NamespacedName` at the beginning of the function so that we can skip for other mutatingwebhookconfigs.

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 14 above)
Comment (ritazh) [reply pkg/reconciler/mutatingwebhook/reconcile.go:69]: If there is an error, then we should return the error so that it can be retried. 
```go
return ctrl.Result{}, client.IgnoreNotFound(err)
```

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 13 above)
Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/reconcile.go:45]: okay makes sense, let me update this. I will use `req.Name` though, since there's no namespace for mutatingwebhookconfigurations

Comment (ritazh) [reply pkg/reconciler/mutatingwebhook/reconcile.go:45]: SG! Both should be fine.

File: pkg/reconciler/mutatingwebhook/reconcile.go (hunk 14 above)
Comment (ritazh) [reply pkg/reconciler/mutatingwebhook/reconcile.go:69]: Do we also need to check if the secret is deleted?

Comment (snehachhabria) [reply pkg/reconciler/mutatingwebhook/reconcile.go:69]: I've updated the retry logic when the update fails. 

As of the secret being deleted, the GetCertificate handles that already so I think we should be okay

**Review by shashankram — APPROVED** [review]

**Review by ritazh — APPROVED** [review]
lgtm
